### PR TITLE
Vendoring in libnetwork to fix #13873.

### LINKED ISSUE #13873: 1.7.1-rc1/2 - container cannot connect establish a tcp socket to itself via the hosts IP address / this works in 1.6.1
Problem
=======
I'm trying to model a distributed system (MongoDB) with a set of containers. In order to establish a quorum, each process takes a list of nodes and confirms a tcp connection between the nodes.
```
+-- host 1 192.x.x.10- / 28001 -+                      
| +-- container 1 178.x.x.20 / 27017   

+-- host 2 192.x.x.11 / 28002--+
| +-- container 2 178.x.x.21/27017
```
So in the example above container 1 tries to open tcp socket to 192.x.x.10 / 28001 and 192.x.x.11 / 28002, container 2 does the same thing.

In 1.6.1 this succeeds. in 1.7.0-rc1 and rc2 this connection to the IP of the host that the container resides fails.

Reproduce
=========
This is the original testcase
```
docker-machine create -d virtualbox --virtualbox-boot2docker-url https://github.com/tianon/boot2docker/releases/download/v1.7.0-rc2/boot2docker.iso dev17

eval "$(docker-machine env dev17)"

docker run -p 28001:27017 -v /data/db/s1rs1a:/data/db --name s1rs1a -d mongo --replSet rs1 --smallfiles --noprealloc 

docker run -p 28002:27017 -v /data/db/s1rs1b:/data/db --name s1rs1b -d mongo --replSet rs1 --smallfiles --noprealloc

mongo <host>:28001
> use admin
> cfg={_id:"rs1",members:[{_id:0,host:"<host ip>:28001"},{_id:1,host:"<host ip>"}]}
> rs.initiate(cfg)
```

Workaround:
==========
None

### PR DESCRIPTION
**Pull request by mavenugo**
Libnetwork sha# e578e95aa101441481411ff1d620f343895f24fe

Fixes #13873.

Signed-off-by: Madhu Venugopal madhu@docker.com

### CODE REVIEW

**Comment by jfrazelle** [comment]
LGTM

File: integration-cli/docker_cli_nat_test.go [hunk 1]
```diff
@@ -98,3 +98,20 @@ func (s *DockerSuite) TestNetworkLocalhostTCPNat(c *check.C) {
 		c.Fatalf("Expected message %q but received %q", msg, final)
 	}
 }
+
+func (s *DockerSuite) TestNetworkLoopbackNat(c *check.C) {
+	testRequires(c, SameHostDaemon, NativeExecDriver)
+	msg := "it works"
+	startServerContainer(c, msg, 8080)
+	endpoint := getExternalAddress(c)
+	runCmd := exec.Command(dockerBinary, "run", "busybox",
```
Comment (icecrime) [inline integration-cli/docker_cli_nat_test.go]: The test seems to be passing on master, both with and without userland-proxy. Also: aren't we supposed to be testing a container connecting to itself through the host IP?

**Comment by icecrime** [comment]
Not LGTM: we're working with @mavenugo to update the test by having the client container be run with `--net=container:server` to indeed test the use case of a container connecting to itself through the host IP.

**Comment by icecrime** [comment]
LGTM with last update :+1: Thanks @mavenugo!

**Comment by mavenugo** [comment]
@icecrime the tests failed :-( ... looking at it now.

**Comment by icecrime** [comment]
The test failure seems like a client-side race condition in busybox' `nc`. Relevant source is [here](http://git.busybox.net/busybox/tree/networking/nc.c?h=1_19_stable#n269): data is outputted through `xwrite` but never flushed.

**Comment by mavenugo** [comment]
@icecrime am in the same boat. since the connection succeeds, but the data is not flushed, I did a temporary fix to ignore the data comparison and accept the connectivity success as the criteria. This atleast doesnt block us due to possible nc client issues. am looking at alternatives in parallel. Being a test env specific issue, I think we can proceed ahead with the connectivity confirmation.

**Comment by mavenugo** [comment]
@icecrime giving it another go with `stty raw`.

**Comment by mavenugo** [comment]
@icecrime `stty raw` used with nc client did the trick :-)

**Comment by cpuguy83** [comment]
LGTM

**Comment by icecrime** [comment]
Ha! Nice :-)

**Comment by mavenugo** [comment]
@jfazelle @icecrime can you pls cherry-pick this to 1.7 bump ?

**Comment by LK4D4** [comment]
Hehe, we should really write our own nc :)

**Comment by jfrazelle** [comment]
chery-picked

**Pull request merged**